File: app/typescript/src/application/service/ItemService.ts
import { ItemUseCase, CreateItemCommand, UpdateItemCommand } from '../port/in/ItemUseCase';
import { ItemRepository } from '../port/out/ItemRepository';
import { Item } from '../../domain/model/item/Item';
import { ItemNotFoundException, DuplicateItemException } from '../../domain/exception/DomainException';

/**
 * 品目アプリケーションサービス
 */
export class ItemService implements ItemUseCase {
  constructor(private readonly itemRepository: ItemRepository) {}

  async createItem(command: CreateItemCommand): Promise<Item> {
    // 重複チェック
    const existing = await this.itemRepository.findByCode(command.品目コード);
    if (existing) {
      throw new DuplicateItemException(command.品目コード);
    }

    const item = Item.create({
      品目コード: command.品目コード,
      適用開始日: new Date(),
      品名: command.品名,
      品目区分: command.品目区分,
      品目グループコード: command.品目グループコード,
      単位コード: command.単位コード,
      場所コード: command.場所コード,
      リードタイム: command.リードタイム,
      安全在庫数: command.安全在庫数,
    });

    return await this.itemRepository.save(item);
  }

  async updateItem(command: UpdateItemCommand): Promise<Item> {
    const item = await this.itemRepository.findByCode(command.品目コード);
    if (!item) {
      throw new ItemNotFoundException(command.品目コード);
    }

    // 更新処理
    if (command.品名) item.品名 = command.品名;
    if (command.品目区分) item.品目区分 = command.品目区分;
    if (command.リードタイム !== undefined) item.リードタイム = command.リードタイム;
    if (command.安全在庫数 !== undefined) item.安全在庫数 = command.安全在庫数;

    return await this.itemRepository.save(item);
  }

  async getAllItems(): Promise<Item[]> {
    return await this.itemRepository.findAll();
  }

  async getItemByCode(itemCode: string): Promise<Item> {
    const item = await this.itemRepository.findByCode(itemCode);
    if (!item) {
      throw new ItemNotFoundException(itemCode);
    }
    return item;
  }

  async getItemsByCategory(category: string): Promise<Item[]> {
    return await this.itemRepository.findByCategory(category);
  }

  async deleteItem(itemCode: string): Promise<void> {
    const item = await this.itemRepository.findByCode(itemCode);
    if (!item) {
      throw new ItemNotFoundException(itemCode);
    }
    await this.itemRepository.deleteByCode(itemCode);
  }
}

File: app/java/src/main/java/com/example/production/application/service/ItemService.java
package com.example.production.application.service;

import com.example.production.application.port.in.ItemUseCase;
import com.example.production.application.port.in.command.CreateItemCommand;
import com.example.production.application.port.in.command.UpdateItemCommand;
import com.example.production.application.port.out.ItemRepository;
import com.example.production.domain.exception.DuplicateItemException;
import com.example.production.domain.exception.ItemNotFoundException;
import com.example.production.domain.model.item.Item;
import com.example.production.domain.model.item.ItemCategory;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;

/**
 * 品目アプリケーションサービス
 */
@Service
@RequiredArgsConstructor
@Transactional
public class ItemService implements ItemUseCase {

    private final ItemRepository itemRepository;

    @Override
    public Item createItem(CreateItemCommand command) {
        // 重複チェック
        itemRepository.findByItemCode(command.getItemCode())
                .ifPresent(existing -> {
                    throw new DuplicateItemException(command.getItemCode());
                });

        Item item = Item.builder()
                .itemCode(command.getItemCode())
                .effectiveFrom(LocalDate.now())
                .itemName(command.getItemName())
                .itemCategory(command.getCategory())
                .unitCode(command.getUnitCode())
                .leadTime(command.getLeadTime() != null ? command.getLeadTime() : 0)
                .safetyLeadTime(command.getSafetyLeadTime() != null ? command.getSafetyLeadTime() : 0)
                .safetyStock(command.getSafetyStock())
                .yieldRate(command.getYieldRate())
                .minLotSize(command.getMinLotSize())
                .lotIncrement(command.getLotIncrement())
                .maxLotSize(command.getMaxLotSize())
                .shelfLife(command.getShelfLife())
                .build();

        itemRepository.save(item);
        return item;
    }

    @Override
    public Item updateItem(UpdateItemCommand command) {
        Item item = itemRepository.findByItemCode(command.getItemCode())
                .orElseThrow(() -> new ItemNotFoundException(command.getItemCode()));

        // 更新処理（null でない場合のみ更新）
        applyUpdates(item, command);

        itemRepository.update(item);
        return item;
    }

    private void applyUpdates(Item item, UpdateItemCommand command) {
        if (command.getItemName() != null) item.setItemName(command.getItemName());
        if (command.getCategory() != null) item.setItemCategory(command.getCategory());
        if (command.getLeadTime() != null) item.setLeadTime(command.getLeadTime());
        if (command.getSafetyLeadTime() != null) item.setSafetyLeadTime(command.getSafetyLeadTime());
        if (command.getSafetyStock() != null) item.setSafetyStock(command.getSafetyStock());
        if (command.getYieldRate() != null) item.setYieldRate(command.getYieldRate());
        if (command.getMinLotSize() != null) item.setMinLotSize(command.getMinLotSize());
        if (command.getLotIncrement() != null) item.setLotIncrement(command.getLotIncrement());
        if (command.getMaxLotSize() != null) item.setMaxLotSize(command.getMaxLotSize());
        if (command.getShelfLife() != null) item.setShelfLife(command.getShelfLife());
    }

    @Override
    @Transactional(readOnly = true)
    public List<Item> getAllItems() {
        return itemRepository.findAll();
    }

    @Override
    @Transactional(readOnly = true)
    public Item getItemByCode(String itemCode) {
        return itemRepository.findByItemCode(itemCode)
                .orElseThrow(() -> new ItemNotFoundException(itemCode));
    }

    @Override
    @Transactional(readOnly = true)
    public List<Item> getItemsByCategory(ItemCategory category) {
        return itemRepository.findByCategory(category);
    }

    @Override
    @Transactional(readOnly = true)
    public List<Item> getItemsByCodes(List<String> itemCodes) {
        return itemRepository.findByItemCodes(itemCodes);
    }

    @Override
    public void deleteItem(String itemCode) {
        itemRepository.findByItemCode(itemCode)
                .orElseThrow(() -> new ItemNotFoundException(itemCode));

        itemRepository.deleteByItemCode(itemCode);
    }
}
